### PR TITLE
Refactor creating and updating slugs in mutations

### LINKED ISSUE #5253: Migration checking for unique product slugs is slow
When running [0114_auto_20200129_0815.py](https://github.com/mirumee/saleor/blob/a66d99b71fa21381481e6e1d5bfe459580416a1e/saleor/product/migrations/0114_auto_20200129_0815.py) against 20k+ products, this migration will run almost forever.

Instead of what it is doing at the moment, it could simply do this:
```sql
select count(slug), slug from product_product group by slug having count(slug) > 1;
-- Time: 0.021s
```

If duplicates are found, try to enumerate the number of occurrences:
```
A: Slug -> Slug (don't change it)
B: Slug -> Slug-1
C: Slug -> Slug-2
D: Slug -> Slug-3 -> Conflict -> Slug-4

### PR DESCRIPTION
**Pull request by IKarbowiak**
Fixes #5253

### Pull Request Checklist

1. [ ] Privileged views and APIs are guarded by proper permission checks.
1. [ ] All visible strings are translated with proper context.
1. [ ] All data-formatting is locale-aware (dates, numbers, and so on).
1. [ ] Database queries are optimized and the number of queries is constant.
1. [ ] Database migration files are up to date.
1. [ ] The changes are tested.
1. [ ] GraphQL schema and type definitions are up to date.
1. [ ] Changes are mentioned in the changelog.

### CODE REVIEW

**Comment by django-queries** [comment]
Here is the report for a3228f6d9ef787f125dd7a5b17decfd92e79529e (mirumee/saleor @ update-performance-in-migrations-with-slugs)
Base comparison is 1fea8259e0b320555bfcb3992ae103dea67c0e09.



<details><summary>No differences were found. (click me)</summary>
<p>

```diff
# api.benchmark checkout
  test name                                  	left count 	right count	duplicate count
  -------------------------------------------	-----------	-----------	---------------
  add billing address to checkout            	         34	         34	             20
  add shipping to checkout                   	          7	          7	              0
  checkout payment charge                    	         10	         10	              0
  complete checkout                          	          8	          8	              0
  create checkout                            	          5	          5	              1

# api.benchmark homepage
  test name                                  	left count 	right count	duplicate count
  -------------------------------------------	-----------	-----------	---------------
  retrieve main menu                         	          5	          5	              0
  retrieve product list                      	          4	          4	              0
  retrieve secondary menu                    	          5	          5	              0
  retrieve shop                              	          2	          2	              0

# api.benchmark product
  test name                                  	left count 	right count	duplicate count
  -------------------------------------------	-----------	-----------	---------------
  product details                            	         17	         17	              3
  retrieve product attributes                	          9	          9	              0

# api.benchmark variant
  test name                                  	left count 	right count	duplicate count
  -------------------------------------------	-----------	-----------	---------------
  product variant bulk create                	         51	         51	              3
  retrieve variant list                      	         23	         23	              9

# api product sorting attributes
  test name                                  	left count 	right count	duplicate count
  -------------------------------------------	-----------	-----------	---------------
  sort product not having attribute data     	         21	         21	              0
```

</p>
</details>

File: saleor/product/migrations/0112_auto_20200129_0050.py [hunk 1]
```diff
@@ -1,49 +1,64 @@
...
-            product_type.save(update_fields=["slug"])
+    for product_type in ProductType.objects.filter(slug__isnull=True):
+        slug = slugify(product_type.name)
+        extension = 1
+        slug_value = slug
+        while True:
+            try:
+                with transaction.atomic():
+                    product_type.slug = slug_value
+                    product_type.save(update_fields=["slug"])
```
Comment (NyanKiyoshi) [inline saleor/product/migrations/0112_auto_20200129_0050.py]: This transaction is not needed

File: saleor/product/migrations/0112_auto_20200129_0050.py [hunk 2]
```diff
@@ -1,49 +1,64 @@
...
+    models_to_update = ["Category", "Collection"]
 
-    for model in models:
+    for model in models_to_update:
         Model = apps.get_model("product", model)
 
-        for instance in Model.objects.all():
-            instance.slug = generate_unique_slug(instance, instance.slug)
-            instance.save(update_fields=["slug"])
+        model_slugs_count = (
```
Comment (NyanKiyoshi) [inline saleor/product/migrations/0112_auto_20200129_0050.py]:
```suggestion
        duplicated_slugs = (
```

File: saleor/warehouse/migrations/0003_warehouse_slug.py [hunk 3]
```diff
@@ -1,37 +1,26 @@
...
-    return unique_slug
+    for warehouse in Warehouse.objects.filter(slug__isnull=True):
+        slug = slugify(warehouse.name)
+        extension = 1
+        slug_value = slug
+        while True:
+            try:
+                with transaction.atomic():
+                    warehouse.slug = slug_value
+                    warehouse.save(update_fields=["slug"])
```
Comment (NyanKiyoshi) [inline saleor/warehouse/migrations/0003_warehouse_slug.py]: Not needed. Wrap `create_unique_slug_for_warehouses` into an atomic transaction instead

File: saleor/product/migrations/0112_auto_20200129_0050.py (hunk 1 above)
Comment (IKarbowiak) [reply saleor/product/migrations/0112_auto_20200129_0050.py]: I need transcation somewhere, otherwise I get an error when running mutation.

Comment (IKarbowiak) [reply saleor/product/migrations/0112_auto_20200129_0050.py]: I changed soultion.

File: saleor/warehouse/migrations/0003_warehouse_slug.py (hunk 3 above)
Comment (IKarbowiak) [reply saleor/warehouse/migrations/0003_warehouse_slug.py]: I changed soultion, there is no more transaction.

**Pull request merged**